### ISSUE TITLE
Usability improvements to "Set Suggest Proof Using" 

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#16151]
If we make the output from this option a warning, then the file position will appear with the message in the coqc output and sent to IDEs.  In CoqIDE, that would create a tooltip in the script, however tooltips that have multiple lines of text currently don't work (nothing pops up).

```
File "./PffAnnotated.v", line 7650, characters 0-4:
Warning: The proof of errorBoundedMultExp_aux
should start with one of the following commands:
Proof using pGivesBound precisionGreaterThanOne radixMoreThanZERO.
```

If a user wants to review the annotation status of an entire package, they can use `-set "Suggest Proof Using"` on the command line.  Since these can be numerous, it may be helpful to have another command line option that prints a single line per file showing the number of messages that `Suggest Proof Using` would generate.

The annotation process seems tedious but hard to support better.  Would it be worth the effort to provide a way to automatically modify the .v file (assuming it's not practical to eliminate the need for annotations)?

I suppose that using more specific annotations rather than `All`, `Type` or `Type*` is generally better for async performance because the latter will require additional sections to complete.  If that's correct, maybe worth mentioning in the manual.

```
The proof of Zpower_nat_monotone_S
should start with one of the following commands:
Proof using nMoreThanOne. 
Proof using All. 
Proof using Type*.
```

If these make sense, I'm willing to work on some of these items.

**Comment by Blaisorblade** [comment]
Improvements are sorely needed, but if we have code that automatically modifies .v files, it should ideally not live in CoqIDE but be shared across editors. I'm also unsure of a good strategy for automatic patching.

My pet peeve: those annotations are IME _always_ wrong. IIRC:
1. Even if I've set `Set Default Suggest ...`, Coq's suggestion will repeat the assumptions that are already implied by the setting. Instead, I only ever want to name the extra assumptions.
2. You can add assumptions to the context without naming them via `Context` — and that's fine — but if those are needed, Coq will sometimes suggest using generated names, and that's always sloppy and sometimes forbidden by project CI.

> I suppose that using more specific annotations rather than All, Type or Type* is generally better for async performance because the latter will require additional sections to complete.

I don't follow what you mean about performance — extra assumptions will change the public API and will make the theorem strictly weaker. OTOH listing assumptions by hand does not always scale. For instance, the following code
```
(** Consequences of the [FinMap] laws. *)
Section fin_maps_lemmas.
  Context `{FinMap K M}.

  (* ... *)
End fin_maps_lemmas.
```
adds to the context 9 TC instances with auto-generated names, and I wouldn't want to list them by hand. In this example, what I'd probably do:
```
Section fin_maps_lemmas.
  Context `{FM : FinMap K M}.
  Local Set Default Proof Using "FM".

  (* ... *)
End fin_maps_lemmas.
```

**Comment by jfehrle** [comment]
@mrhaandi I made an unstated supposition that may be wrong: that each section variable in a `Proof using` is an additional dependency that needs to be satisfied before the proof can be processed.  I expect the best practice is using the more precise annotation whether my supposition is correct or not?  If you can confirm the reason, I'll add that to the doc.

> From my experience adapting half a thousand files to `Proof using`, setting up a batch `Suggest Proof Using` will just pollute the output.

Because processing many files, the number of messages would be overwhelming?

> What would have helped me, is batch output (file, line number) only suggestions for lemmas where `Proof using Type` would fail.

Is this a general purpose or specific to your project(s)?  Was your strategy to add `Proof using Type` _en masse_ as the default annotation?

**Comment by Blaisorblade** [comment]
> @mrhaandi I made an unstated supposition that may be wrong: that each section variable in a Proof using is an additional dependency that needs to be satisfied before the proof can be processed.

`using` only affects which section variables can be actually used in the proof; `Qed` will reject proofs that depend on extra assumptions, and closing a section will add the assumptions listed by `using` as hypotheses of the theorem (but not other section variables).

> I expect the best practice is using the more precise annotation whether my supposition is correct or not?

That's at least the baseline recommendation (my previous message tried to describe additional complications).

**Comment by jfehrle** [comment]
@Blaisorblade A quick response on the first part of your comment:

> Improvements are sorely needed, but if we have code that automatically modifies .v files, it should ideally not live in CoqIDE but be shared across editors. I'm also unsure of a good strategy for automatic patching.

We certainly don't want to tie a generally useful tool to a single IDE.  It could be an executable or available through an RPC.  Even something quick and dirty that makes some mistakes could be helpful.  Inventing refactoring infrastructure would be non-trivial.

**Comment by mrhaandi** [comment]
> Is this a general purpose approach or specific to your project(s)? Was your strategy to add `Proof using Type` _en masse_ as the default annotation?

Yes, this seems to me as the only manual option which scales. In my use case 95% of lemmas were happy with `Proof using Type`, so I just need to know about the remaining 5%.
The original code is quite heterogenous, with a dozen developers and several styles involved.

> Because processing many files, the number of messages would be overwhelming?

Many files, a lot of which would be happy with `Proof using Type` entirely.
Many lemmas per file, with only a few which need a different `using` annotation.
An unfiltered `Suggest Proof Using` output would amount to tens of thousands of suggestion, with only a hundred important ones.

**Comment by Blaisorblade** [comment]
@mrhaandi How does that compare with `#[local] Set Default Proof Using "Type"` at the beginning of each file + fixing errors?

**Comment by mrhaandi** [comment]
> @mrhaandi How does that compare with `#[local] Set Default Proof Using "Type"` at the beginning of each file + fixing errors?

What I ended up doing, which seems least laborious, is globally `-set Default Proof Using "Type"` and manually fix batch build errors, which provide most crucial information.

**Comment by jfehrle** [comment]
How about this instead?  I could add a CLI option to `coqc` to add or update all the `Proof using` commands using the first suggestion from `Suggest Proof Using`.  That may only take a day or two to implement.  Files would be updated in-place.  You could use a makefile to update an entire project.  (Supporting something like this  in the IDE environment maybe much more difficult because the user may be editing the file and they may backtrack in the proof.)

**Comment by jfehrle** [comment]
> @Blaisorblade 1. Even if I've set `Set Default Suggest ...`, Coq's suggestion will repeat the assumptions that are already implied by the setting. Instead, I only ever want to name the extra assumptions.

I don't see any such option.  If it did, I can see your point.

> 2. You can add assumptions to the context without naming them via `Context` — and that's fine — but if those are needed, Coq will sometimes suggest using generated names, and that's always sloppy and sometimes forbidden by project CI.

The documentation for `Context` seems woefully incomplete.  Checking for generated names in the automated tool I suggested above and then adding/correcting `Context` statements is likely more work than I'd want to take on.  The sloppiness would be uncovered by error messages over time as the variable names change or when you run CI.  Is the CI check generally available or specific to your environment?

**Comment by gares** [comment]
Yes, as I wrote in of the the many threads about this topic the .aux files almost contain the right piece of info. I think the only problem is that the location the carry is the one of `Qed`, not the one of `Proof` which would be ideal. So one has to walk back the file and find (or add) the `Proof` corresponding to the suggestion.

**Comment by jfehrle** [comment]
I did something simpler: My tool inserts the first suggestion from `Suggest Print Using`.  After updating my file with my tool, I see these problems:

1 I get a few errors like this (using coqc and `-async-proofs off`).  These suggests some bugs in managing parallel tasks
@gares is this a known issue or something worth reducing to a short example?

```
File "./Pff.v", line 4130, characters 0-4:
Error:  (in proof FPredDiff1): Attempt to save an incomplete proof
```

2. The annotated file takes 4:23 in CoqIDE with 1 async job, 3:00 with 2 jobs and 2.45 with 4 jobs.  I can't run the annotated file with async off, but if I run the original file it takes 1:34.  Total CPU time ~2 mins for async off and ~7 minutes for 4 cores.

3. The number of jobs visible in the GUI didn't match always match the requested number (sometimes more and sometimes less) and they often reported Idle status.  This might just be from delayed messages.

**Comment by gares** [comment]
1. if you get the same error with async off, how can it be a bug in managing parallel tasks?

2. I'm starting to get bored of repeating this: you are measuring the wrong thing. Start CoqIDE, start a clock, jump to the end of the file, when the "coq busy" widget stops look at the clock. With the proof using annotations that should be faster, as well as .vio and .vos generation in batch mode. Lowering the time to completely check the file is not what this whole machinery is for. (Jumping to the end of the file is just to ease the measurement, it should be faster to jump forward anywhere in the file)

3. this is very possible, I think I did set a timer which updates the panel from time to time

**Comment by jfehrle** [comment]
> if you get the same error with async off, how can it be a bug in managing parallel tasks?

Yeah, my mistake.  It appears to be caused by an incorrect annotation on an earlier proof that generates no error message.

> I'm starting to get bored of repeating this: you are measuring the wrong thing. Start CoqIDE, start a clock, jump to the end of the file, when the "coq busy" widget stops look at the clock.

That's what I'm doing, that's what I've been doing all along: timing it with my phone, waiting until the last lines turn green.  Perhaps there is something about WSL+Ubuntu that impacts the scaling.

**Comment by jfehrle** [comment]
> The annotated file takes 4:23 in CoqIDE with 1 async job, 3:00 with 2 jobs and 2.45 with 4 jobs. I can't run the annotated file with async off, but if I run the original file it takes 1:34.  Total CPU time ~2 mins for async off and ~7 minutes for 4 cores.

The first sentence is measured with my phone.  When I say "CPU time", that's from running CoqIDE in the `time` command--I've only done that a couple times, the main thing is with my phone.

**Comment by gares** [comment]
> But FYI that turns green well before the run completes.

This is what I'm trying to say, not clearly enough. The whole point of the async mode is that Coq can respond to user request sooner, way before having completed the check of the document.

Instead of `Check 1` write a partial, unfinished, proof script you are working on. You figure out you need to change a lemma up in the file, you go there, change it, and go back working on your last proof. This is what async optimizes. *NOT* the full checking of the document.
